clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -983,7 +984,7 @@ private ClusterAndWaitTime waitOnMetadata(String topic, Integer partition, long
         } while (partitionsCount == null);
 
         if (partition != null && partition >= partitionsCount) {
-            throw new KafkaException(
+            throw new UnknownTopicOrPartitionException(
```

Comment:
should we include the topic name and partition number in this exception, so the calling code can check which topic or partition is missing?

```
try {
  producer.send(.. args ..);
} catch (UnknownTopicOrPartitionException e) {
  log.error("missing topic {} partition {}", e.topic(), e.partition())
}
```
also in this case we are very sure that the topic exists, but the partition does not. do you think we should add a `UnknownPartitionException` here?

Reply:
I think the answer to both questions depends on how you imagine clients using this. Since it's a retriable exception, would the client just log the information and retry, or are there particular actions to take based on the topic or partition? If all we care about is logging, the exception message already includes the topic and partition, but if there's a use case for adding the topic and partition to the exception then I agree that we should do that.

Reply:
Ah, didn't realize this was a `RetriableException`. we might then want to add what topic caused the problem too. Otherwise, if the topic is not being created, clients (for example, Connect  would be stuck in a retry loop forever without any guidance on what topic needs to be fixed).

P.S: hopefully, we aren't skipping writing a KIP here since this does change the behavior of a public API.